src/libeconf.c

```diff
@@ -133,3 +133,71 @@ void merge_files(const char *save_to_dir, const char *file_name,
   destroy(etc_file);
   destroy_merged_file(merged_file);
 }
+
+/* GETTER FUNCTIONS */
+// TODO: Return values in error case
+
+// TODO: Currently only works with a sorted Key_File. If a new
+// key with an existing group is appended at the end the group
+// will show twice. So the key file either needs to be sorted
+// upon entering a new key or the function must ensure only
+// unique values are returned.
+char **getGroups(Key_File kf, size_t *length) {
+  *length = 0;
+  char *uniques = malloc(kf.length + 1);
+  char *ptr = uniques;
+  for (int i = 0; i < kf.length; i++) {
+    if (!i || strcmp(kf.file_entry[i].group, kf.file_entry[i-1].group)) {
+      *uniques++ = '1';
+      (*length)++;
+      continue;
+    }
+    *uniques++ = '0';
+  }
+  *uniques = 0;
+  uniques = ptr;
+  size_t num = 0;
+  char **groups = malloc(*length * sizeof(char*));
+  *length = 0;
+  while(*uniques) {
+    if (*uniques == '1') {
+      num = strlen(kf.file_entry[uniques - ptr].group) + 1;
+      groups[*length] = malloc(num);
+      snprintf(groups[(*length)++], num, kf.file_entry[uniques - ptr].group);
+    }
+    *uniques++;
+  }
+  free(ptr);
+  return groups;
+}
+
+// TODO: Same issue as with getGroups()
+char **getKeys(Key_File kf, const char *group, size_t *length) {
+  *length = 0;
+  char *uniques = malloc(kf.length + 1);
+  char *ptr = uniques;
+  for (int i = 0; i < kf.length; i++) {
+    if (!strcmp(kf.file_entry[i].group, group) &&
+        (!i || strcmp(kf.file_entry[i].key, kf.file_entry[i-1].key))) {
+      *uniques++ = '1';
+      (*length)++;
+      continue;
+    }
+    *uniques++ = '0';
+  }
+  *uniques = 0;
+  uniques = ptr;
+  size_t num = 0;
+  char **keys = malloc(*length * sizeof(char*));
+  *length = 0;
+  while(*uniques) {
+    if (*uniques == '1') {
+      num = strlen(kf.file_entry[uniques - ptr].key) + 1;
+      keys[*length] = malloc(num);
+      snprintf(keys[(*length)++], num, kf.file_entry[uniques - ptr].key);
+    }
+    *uniques++;
+  }
+  free(ptr);
+  return keys;
+}
```

Comment:
Suggestion: Use iterator-like pattern - still only works with sorted Key_Files, but it may be possible to compare with previous entries (O(n²))